algorithm/cgal/cgal.cc

```diff
@@ -1205,6 +1272,112 @@ const Surface_mesh* ApproximateSurfaceMesh(
   return output.release();
 }
 
+template <typename Vertex_point_map>
+bool is_coplanar_edge(const Surface_mesh& m, const Vertex_point_map& p,
+                      const Halfedge_index e) {
+  auto a = p[m.source(e)];
+  auto b = p[m.target(e)];
+  auto c = p[m.target(m.next(e))];
+  typename Kernel::Plane_3 plane(a, b, c);
+  return plane.has_on(p[m.target(m.next(m.opposite(e)))]);
+}
+
+template <typename Vertex_point_map>
+bool is_collinear_edge(const Surface_mesh& m, const Vertex_point_map& p,
+                       const Halfedge_index e0, const Halfedge_index e1) {
+  // Assume that e0 and e1 share the same source vertex.
+  const auto& a = p[m.source(e0)];
+  const auto& b = p[m.target(e0)];
+  const auto& c = p[m.target(e1)];
+  return CGAL::collinear(a, b, c);
+}
+
+template <typename Vertex_point_map>
+bool is_safe_to_move(const Surface_mesh& m, const Vertex_point_map& p,
+                     const Halfedge_index start) {
+  for (Halfedge_index e = m.next_around_source(start); e != start;
+       e = m.next_around_source(e)) {
+    if (!is_coplanar_edge(m, p, e) && !is_collinear_edge(m, p, start, e)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename Kernel>
+class Demesh_cost {
```

Comment:
How did the cost come out? Is it expensive?

Reply:
The cost is used to prioritize the edge collapses.